### PR TITLE
Cleanup and refactoring

### PR DESCRIPTION
**Pull request by devanych**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | -

### CODE REVIEW

**Comment by devanych** [comment]
> .



> Need overall benchmark to ensure logger doesn't consume CPU/memory more than it was before refactoring. That is critical framework component in this regard.

There is one caveat that checks were added here, but they are not in the master. For true benchmarking, must also add checks to the master.

**Comment by samdark** [comment]
Benchmark doesn't reveal anything critical about refactoring.

**Comment by samdark** [comment]
@devanych would you please resolve conflicts?

File: src/Logger.php [hunk 1]
```diff
@@ -109,16 +105,30 @@ public function __construct(array $targets = [])
...
-        if (is_string($level)) {
-            return $level;
+        if (!is_string($level)) {
+            throw new InvalidArgumentException(sprintf(
+                'The log message level must be a string, %s received.',
+                gettype($level)
+            ));
+        }
+
+        $level = strtolower($level);
```
Comment (samdark) [inline src/Logger.php:121]: Do we really need to support non-lowercased levels? I think we can remove it.

File: src/MessageCategory.php [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+use function is_string;
+use function rtrim;
+use function substr_compare;
+use function sprintf;
+use function strpos;
+
+/**
+ * Category is a data object that stores and matches the included and excluded categories of log messages.
+ */
+final class MessageCategory
```
Comment (samdark) [inline src/MessageCategory.php:19]: I have an idea of a better name: `MessageCategoryFilter`.

File: src/Target.php [hunk 3]
```diff
@@ -92,292 +50,332 @@ abstract class Target
...
-                    $context,
-                    [
-                        'category' => static::DEFAULT_CATEGORY,
-                        'time' => $_SERVER['REQUEST_TIME_FLOAT'] ?? \microtime(true),
-                    ],
-                ];
+            if (($contextMessage = $this->getContextMessage()) !== '') {
+                $this->messages->add(LogLevel::INFO, $contextMessage, [
+                    'category' => MessageCategory::DEFAULT,
+                    'time' => $_SERVER['REQUEST_TIME_FLOAT'] ?? microtime(true),
```
Comment (samdark) [inline src/Target.php:122]: That looks suspicious. ` $_SERVER['REQUEST_TIME_FLOAT']` is the time of the request beginning.

**Review by samdark — CHANGES_REQUESTED** [review]

File: src/Logger.php (hunk 1 above)
Comment (devanych) [reply src/Logger.php:121]: Ok.

File: src/Target.php (hunk 3 above)
Comment (devanych) [reply src/Target.php:122]: Oops! Forgot to delete.

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
👍

**Pull request merged**